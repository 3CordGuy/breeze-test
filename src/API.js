const getGroups = () =>
    fetch(`//localhost:8000/api/groups/`).then((response) => response.json());

const getPeople = () =>
    fetch(`//localhost:8000/api/people/`).then((response) => response.json());

const removePerson = (person_id) =>
    fetch(`//localhost:8000/api/people/${person_id}`, {
        method: "DELETE",
        headers: {
            "Content-Type": "application/json",
        },
    });

const updatePerson = (person) =>
    fetch(`//localhost:8000/api/people/${person.id}`, {
        method: "PUT",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(person),
    });

const importPeople = (people) =>
    fetch(`//localhost:8000/api/people-import`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(people),
    });

const importGroups = (groups) =>
    fetch(`//localhost:8000/api/groups-import`, {
        method: "POST",
        mode: "cors",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(groups),
    });

const getGroup = (group_id) =>
    fetch(`http://localhost:8000/api/groups/${group_id}`).then((response) =>
        response.json(),
    );

const API = {
    getGroups,
    getPeople,
    removePerson,
    updatePerson,
    importPeople,
    importGroups,
    getGroup,
};

export default API;
