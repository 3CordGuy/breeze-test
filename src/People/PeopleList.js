import React, { Component } from "react";
import {
    Table,
    Dimmer,
    Loader,
    Header,
    Icon,
    Button,
    Dropdown,
} from "semantic-ui-react";
import { Link, Location } from "@reach/router";
import API from "../API";
import ImportCSV from "../ImportCSV";
import _ from "lodash";

class PeopleList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            column: "last_name",
            direction: "ascending",
            data: [],
            groups: [],
            loading: true,
        };
    }

    getPeople = () => {
        this.setState({ loading: true });
        API.getPeople().then(({ data }) =>
            this.setState({
                data: _.sortBy(data, ["last_name", "ascending"]),
                loading: false,
            }),
        );
    };

    getGroups = () => {
        this.setState({ loading: true });
        API.getGroups().then(({ data }) =>
            this.setState({
                groups: data.map((group) => ({
                    key: group.id,
                    value: group.id,
                    text: group.name,
                })),
                loading: false,
            }),
        );
    };

    componentDidMount() {
        this.getPeople();
        this.getGroups();
    }

    handleSort = (clicked_column) => () => {
        const { column, data, direction } = this.state;
        if (column !== clicked_column) {
            this.setState({
                column: clicked_column,
                data: _.sortBy(data, [clicked_column]),
                direction: "descending",
            });

            return;
        }

        this.setState({
            data: data.reverse(),
            direction: direction === "ascending" ? "descending" : "ascending",
        });
    };

    handleAssignGroup = (person, group_id) => {
        person.group_id = group_id;
        delete person.group;

        this.setState({ loading: true });
        API.updatePerson(person).then(() => this.getPeople());
    };

    handleRemove = (person) => {
        if (
            window.confirm(
                `Are you sure you want to delete ${person.first_name} ${person.last_name}?!`,
            )
        ) {
            this.setState({ loading: true });
            API.removePerson(person.id).then(() => this.getPeople());
        }
    };

    render() {
        let data = this.state.data || [];
        let { loading, groups, column, direction } = this.state;

        return (
            <>
                {loading && (
                    <Dimmer active inverted>
                        <Loader>Loading Data...</Loader>
                    </Dimmer>
                )}
                <Header as="h1" dividing>
                    <Icon name="user" />

                    <Header.Content as="div" style={{ width: "100%" }}>
                        People
                        <Location>
                            {({ location }) => (
                                <ImportCSV
                                    text="Upload CSV"
                                    location={location}
                                    onFinish={this.getPeople}
                                />
                            )}
                        </Location>
                    </Header.Content>
                </Header>

                <Table celled padded sortable basic="very">
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell
                                singleLine
                                sorted={
                                    column === "first_name" ? direction : null
                                }
                                onClick={this.handleSort("first_name")}
                            >
                                First Name
                            </Table.HeaderCell>
                            <Table.HeaderCell
                                sorted={
                                    column === "last_name" ? direction : null
                                }
                                onClick={this.handleSort("last_name")}
                            >
                                Last Name
                            </Table.HeaderCell>
                            <Table.HeaderCell
                                sorted={
                                    column === "email_address"
                                        ? direction
                                        : null
                                }
                                onClick={this.handleSort("email_address")}
                            >
                                Email
                            </Table.HeaderCell>
                            <Table.HeaderCell
                                sorted={
                                    column === "group.name" ? direction : null
                                }
                                onClick={this.handleSort("group.name")}
                            >
                                Group
                            </Table.HeaderCell>
                            <Table.HeaderCell></Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>

                    <Table.Body>
                        {data.length ? (
                            data.map((person, index) => {
                                return (
                                    <Table.Row key={index}>
                                        <Table.Cell singleLine>
                                            {person.first_name}
                                        </Table.Cell>
                                        <Table.Cell singleLine>
                                            {person.last_name}
                                        </Table.Cell>
                                        <Table.Cell singleLine>
                                            {person.email_address}
                                        </Table.Cell>
                                        <Table.Cell singleLine>
                                            {person.group ? (
                                                <Link
                                                    to={`/groups/${person.group.id}`}
                                                >
                                                    {person.group.name}
                                                </Link>
                                            ) : (
                                                <Dropdown
                                                    selection
                                                    placeholder="Add to Group"
                                                    renderLabel={({ text }) =>
                                                        text
                                                    }
                                                    onChange={(e, data) =>
                                                        this.handleAssignGroup(
                                                            person,
                                                            data.value,
                                                        )
                                                    }
                                                    options={groups}
                                                ></Dropdown>
                                            )}
                                        </Table.Cell>

                                        <Table.Cell
                                            singleLine
                                            textAlign="center"
                                        >
                                            <Button
                                                icon
                                                circular
                                                negative
                                                basic
                                                onClick={() =>
                                                    this.handleRemove(person)
                                                }
                                            >
                                                <Icon name="trash" />
                                            </Button>
                                        </Table.Cell>
                                    </Table.Row>
                                );
                            })
                        ) : (
                            <Table.Row>
                                <Table.Cell textAlign="center">
                                    There are no People yet!{" "}
                                    <span
                                        role="img"
                                        aria-label="sad face"
                                        style={{ fontSize: "2rem" }}
                                    >
                                        üôÅ
                                    </span>
                                </Table.Cell>
                                <Table.Cell />
                            </Table.Row>
                        )}
                    </Table.Body>
                </Table>
            </>
        );
    }
}

export default PeopleList;
