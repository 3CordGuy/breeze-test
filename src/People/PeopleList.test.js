import React from "react";
import { shallow, mount } from "enzyme";
import PeopleList from "./PeopleList";
import API from "../API";
import _ from "lodash";

let PEOPLE_LIST_RESPONSE = {
    data: [
        {
            id: 123,
            first_name: "Leslie",
            last_name: "Knope",
            email_address: "leslise@pawnee.in.gov",
            status: "active",
            created_at: "2020-06-28T01:48:45.000000Z",
            group: {
                id: 63,
                name: "Dunder Mifflin Fan Club",
                created_at: "2020-06-28T01:17:03.000000Z",
                updated_at: "2020-06-28T01:17:03.000000Z",
            },
            updated_at: "2020-06-28T01:48:45.000000Z",
        },
        {
            id: 125,
            first_name: "Fred",
            last_name: "Flintstone",
            email_address: "fredflintstone@example.com",
            status: "active",
            created_at: "2020-06-28T01:48:45.000000Z",
            group: {
                id: 59,
                name: "Volunteers",
                created_at: "2020-06-28T01:17:03.000000Z",
                updated_at: "2020-06-28T01:17:03.000000Z",
            },
            updated_at: "2020-06-28T01:49:08.000000Z",
        },
        {
            id: 126,
            first_name: "Marie",
            last_name: "Bourne",
            email_address: "mbourne@example.com",
            status: "active",
            created_at: "2020-06-28T01:48:45.000000Z",
            group: {
                id: 63,
                name: "Dunder Mifflin Fan Club",
                created_at: "2020-06-28T01:17:03.000000Z",
                updated_at: "2020-06-28T01:17:03.000000Z",
            },
            updated_at: "2020-06-28T01:48:45.000000Z",
        },
        {
            id: 127,
            first_name: "Wilma",
            last_name: "Flintstone",
            email_address: "wilmaflinstone@example.com",
            status: "active",
            created_at: "2020-06-28T01:48:45.000000Z",
            group: {
                id: 61,
                name: "Bible Study",
                created_at: "2020-06-28T01:17:03.000000Z",
                updated_at: "2020-06-28T01:17:03.000000Z",
            },
            updated_at: "2020-06-28T01:49:11.000000Z",
        },
    ],
    total: 4,
};
let wrapper, data;

// SNAPSHOT
describe("<PeopleList />", () => {
    beforeAll(() => {
        wrapper = shallow(<PeopleList />);
        data = [
            {
                id: 132,
                first_name: "Macie",
                last_name: "Emmerich",
                email_address: "cremin.marjory@hotmail.com",
                status: "active",
                updated_at: "2019-07-20 22:05:47",
                created_at: "2019-07-20 22:05:47",
            },
        ];

        wrapper.setState({ data: data });
    });

    test("should match the snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });
});

// UNIT TESTS
describe("<PeopleList />", () => {
    it("should show a proper loader", () => {
        const wrapper = mount(<PeopleList />);
        wrapper.setState({ loading: true });

        let loader = wrapper.find("div.loader");

        expect(loader.text()).toBe("Loading Data...");
    });

    it("should warn when trying to delete a person", () => {
        const wrapper = mount(<PeopleList />);
        window.confirm = jest.fn();
        wrapper.setState({
            data: [
                {
                    id: 132,
                    first_name: "Jim",
                    last_name: "Halpert",
                    email_address: "jim.halpert@dundermifflin.com",
                    status: "active",
                    updated_at: "2019-07-20 22:05:47",
                    created_at: "2019-07-20 22:05:47",
                },
            ],
        });

        let trashButton = wrapper.find("button.negative");
        trashButton.simulate("click");

        expect(window.confirm).toBeCalled();
    });

    it("should fetch the people async", () => {
        const promise = new Promise((resolve) => resolve(PEOPLE_LIST_RESPONSE));

        API.getPeople = jest.fn(() => promise);

        const wrapper = mount(<PeopleList />);

        expect(wrapper.find("table tbody").length).toEqual(1);

        promise.then(() => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find("table tbody tr").length).toEqual(4);

                API.getPeople.mockClear();
            });
        });
    });

    it("should sort list by person last_name", () => {
        const componentWrapper = mount(<PeopleList />);
        const data = []
            .concat(PEOPLE_LIST_RESPONSE.data[0])
            .concat(PEOPLE_LIST_RESPONSE.data[1]);

        componentWrapper.setState({
            column: "last_name",
            direction: "asc",
            data: _.orderBy(data, "last_name", ["asc"]),
        });

        expect(componentWrapper.state().data).toEqual(
            _.orderBy(data, "last_name", ["asc"]),
        );
    });

    it("should sort list by person last_name desc when clicked", () => {
        const componentWrapper = mount(<PeopleList />);
        const data = []
            .concat(PEOPLE_LIST_RESPONSE.data[0])
            .concat(PEOPLE_LIST_RESPONSE.data[1]);

        componentWrapper.setState({
            column: "last_name",
            direction: "asc",
            data: _.orderBy(data, "last_name", ["asc"]),
        });

        let nameHeader = componentWrapper.find("th.ascending");

        nameHeader.simulate("click");

        expect(componentWrapper.state().data).toEqual(data);
    });
});
