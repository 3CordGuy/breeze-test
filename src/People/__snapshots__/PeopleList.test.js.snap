// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PeopleList /> should match the snapshot 1`] = `
<Fragment>
  <Dimmer
    active={true}
    inverted={true}
  >
    <Loader>
      Loading Data...
    </Loader>
  </Dimmer>
  <Header
    as="h1"
    dividing={true}
  >
    <Icon
      as="i"
      name="user"
    />
    <HeaderContent
      as="div"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      People
      <Location>
        <Component />
      </Location>
    </HeaderContent>
  </Header>
  <Table
    as="table"
    basic="very"
    celled={true}
    padded={true}
    sortable={true}
  >
    <TableHeader
      as="thead"
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          singleLine={true}
          sorted={null}
        >
          First Name
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          sorted="ascending"
        >
          Last Name
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          sorted={null}
        >
          Email
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          sorted={null}
        >
          Group
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
        />
      </TableRow>
    </TableHeader>
    <TableBody
      as="tbody"
    >
      <TableRow
        as="tr"
        cellAs="td"
        key="0"
      >
        <TableCell
          as="td"
          singleLine={true}
        >
          Macie
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          Emmerich
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          cremin.marjory@hotmail.com
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        >
          <Dropdown
            additionLabel="Add "
            additionPosition="top"
            closeOnBlur={true}
            closeOnEscape={true}
            deburr={false}
            icon="dropdown"
            minCharacters={1}
            noResultsMessage="No results found."
            onChange={[Function]}
            openOnFocus={true}
            options={Array []}
            placeholder="Add to Group"
            renderLabel={[Function]}
            searchInput="text"
            selectOnBlur={true}
            selectOnNavigation={true}
            selection={true}
            wrapSelection={true}
          />
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
          textAlign="center"
        >
          <Button
            as="button"
            basic={true}
            circular={true}
            icon={true}
            negative={true}
            onClick={[Function]}
          >
            <Icon
              as="i"
              name="trash"
            />
          </Button>
        </TableCell>
      </TableRow>
    </TableBody>
  </Table>
</Fragment>
`;
